### ISSUE TITLE
Button state Controller

### ISSUE THREAD

**Issue by kubapilch** [kivymd/KivyMD#1418]
### Description of the Feature

I have been working with a lot of projects where I use kivyMD and in most of them I have to disable certain buttons depending on the user input/hardware status. When it is a couple of buttons and not that many things/places that should turn them on/off it is manageable, but in bigger projects with many variables for a button to be enabled/disabled it is very hard to keep track of it nicely.

That is why I have written a `ButtonStateController`, its main purpose is to basically streamline enabling/disabling button with all the factors that have to be taken into account.

Let me know if you think that it is something that you would want to add intro the library and if so I can create a pull request. 

### Code and Logs

```python
class IncorrectFactor(Exception):
    pass


class ButtonNotBinded(Exception):
    pass


class ButtonStateController:
    def __init__(
        self,
        factors: dict,
    ) -> None:
        self._factors = factors
        self._factors_binds = {factor: [] for factor in self._factors}
        self._button_objects = {}
        self._button_binds = {}

    def get_factors_names(self) -> tuple:
        """Returns a tuple of available factors

        Returns
        -------
        tuple
            Tuple of available factors
        """
        return self._factors.keys()

    def get_factors(self) -> dict:
        """Returns a dict of factors with current states

        Returns
        -------
        dict
            Factors
        """
        return self._factors

    def get_button_state(
        self, button: Union[MDRectangleFlatButton, MDRectangleFlatIconButton]
    ) -> bool:
        """Returns the current state of a button

        Parameters
        ----------
        button : Union[MDRectangleFlatButton, MDRectangleFlatIconButton]
            Button object

        Returns
        -------
        bool
            Whether the button is disabled on enabled

        Raises
        ------
        ButtonNotBinded
            Exception when the passed button is not binded
        """
        try:
            return self._button_binds[button]["button_state"]
        except KeyError as k:
            raise ButtonNotBinded(k)

    def bind_button(
        self,
        button: Union[MDRectangleFlatButton, MDRectangleFlatIconButton],
        factors: tuple,
        compare_func: Callable[[bool], bool],
    ) -> None:
        """Binds a button to specified factors

        Parameters
        ----------
        button : Union[MDRectangleFlatButton, MDRectangleFlatIconButton]
            Button object
        factors : tuple
            List of factors for the button to react to
        compare_func : Callable
            Function (preferably lambda) to which factors will be passed in the same order they are specified in factors argument list, returning boolean value indicating whether button should be enabled or disabled for that combination of factor values

        Raises
        ------
        IncorrectFactor
            Exception when an incorrect Factor is passed
        """
        try:
            self._button_binds[button] = {
                "button_state": compare_func(*map(self._factors.__getitem__, factors)),
                "factors": factors,
                "compare_func": compare_func,
            }
        except KeyError as k:
            raise IncorrectFactor(f"{k}, valid factors: {self.get_factors_names()}")

        for factor in factors:
            self._factors_binds[factor].append(button)

        if self._button_binds[button]["button_state"]:
            button.disabled = False
        else:
            button.disabled = True

    def set_factor(self, factor: str, state: bool) -> None:
        """Sets a factor and computes new states for affected buttons

        Parameters
        ----------
        factor : str
            Factor name that has to be set
        state : bool
            New state of the factor

        Raises
        ------
        IncorrectFactor
            Exception when an incorrect Factor is passed
        """
        try:
            if self._factors[factor] == state:
                return

            self._factors[factor] = state
        except KeyError as k:
            raise IncorrectFactor(k)

        for button in self._factors_binds[factor]:
            desired_state = self._button_binds[button]["compare_func"](
                *map(self._factors.__getitem__, self._button_binds[button]["factors"])
            )

            if desired_state == self._button_binds[button]["button_state"]:
                continue

            if desired_state:
                button.disabled = False
            else:
                button.disabled = True

            self._button_binds[button]["button_state"] = desired_state
```

#### Creating an object
```python
self.bc = ButtonStateController(
            {
                "idle": True,
                "measuring": False,
            }
        )
```

#### Binding buttons
This action will bind button with ID 'stop_measurment' to two factors and evaluate button state with lambda function
```python
self.bc.bind_button(
            self.root.ids.stop_measurment, ("idle", "measuring"), lambda i, m: not i and m
        )
```

#### Changing factor state
This action will update factor 'idle' state and iterate over all buttons that are connected to that factor and evaluate their new state
```python
self.bc.set_factor("idle", True)
```

### Versions

* OS: Any
* Python: 3.10
* Kivy: 2.1
* KivyMD: 1.1.1

**Comment by HeaTTheatR** [comment]
@kubapilch I don't understand why this is necessary...

**Comment by kubapilch** [comment]
@HeaTTheatR I am not sure how many people it is affecting. I know from my projects where I have over 15 buttons that need to be dynamically enabled/disabled depending on the state of the machine that the UI is controlling and user interactions that it might get messy.

For some of those buttons, I have multiple criteria that need to be satisfied in order to enable/disable a button. All of this can change in a lot of places in code and while scaling up the project new criteria might be added, in which case I have to go to every place in code when I'm changing button state to add a new if condition. This has led to several bugs with the UI in my team.

I don't know how useful it would be for a typical user of the library, but I have found that since we have implemented that controller, it is much easier to manage button states. As I had the code already written, I created this Issue to ask if you think that it might be something worth adding.
